apps/rush-buildxl/src/api/BxlModulesGenerator.ts

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { RushConfiguration } from '@microsoft/rush-lib';
+import { BxlModule } from './BxlModule';
+import { BxlConfig } from './BxlConfig';
+
+export class BxlModulesGenerator {
+  private _rushConfiguration: RushConfiguration;
+  private _buildXLRoot: string;
+
+  constructor(rushConfiguration: RushConfiguration, buildXLRoot: string) {
+    this._rushConfiguration = rushConfiguration;
+    this._buildXLRoot = this._normalizePathSeparator(buildXLRoot);
+  }
+
+  public run(): boolean {
+    const modulesRoot: string = this._normalizePathSeparator(`${this._rushConfiguration.commonTempFolder}/bxl`);
+    const rushJsonFile: string = this._normalizePathSeparator(this._rushConfiguration.rushJsonFile);
+
+    const modules: BxlModule[] =  this._rushConfiguration.projects.map(project => {
+      const name: string = this._packageNameToModuleName(project.packageName);
+      const moduleRoot: string = `${modulesRoot}/${name}`;
+      const projDir: string = this._normalizePathSeparator(project.projectFolder);
+
+      return new BxlModule(name, projDir, rushJsonFile, moduleRoot);
+    });
+
+    const bxlConfig: BxlConfig = new BxlConfig(this._buildXLRoot, modulesRoot, modules);
+
+    const tasks: Array<Promise<void>> = modules.map(module => module.writeFile());
+    tasks.push(bxlConfig.writeFile());
+
+    let success: boolean = true;
+    Promise
+      .all(tasks)
+      .catch(() => {
+        success = false;
+      });
+
+    return success;
+  }
+
+  private _packageNameToModuleName(packageName: string): string {
+    return packageName.replace('/', '_');
```

Comment:
Are other characters not allowed? Like `@`?

Reply:
Only replacing '/' since it's a path separator and allows modules to be generated in a single directory.